### PR TITLE
`Learning paths`: Add explanation view for learning path users

### PR DESCRIPTION
**Pull request by JohannesWt**
> [!WARNING]  
> ### Contains database migration

#### General

- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a [very good performance](https://docs.artemis.cit.tum.de/dev/guidelines/performance/) and prevented too many (unnecessary) and too complex database calls.
- [x] I **strictly** followed the principle of **data economy** for all database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data (e.g. using paging).
- [x] I **strictly** followed the principle of **data economy** for all client-server REST calls.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

Currently students get directly confronted with the learning path feature without an explanation. This PR introduces a simple explanation and enables students to start their learning path on their own.

### Description

This PR adds a new column in the learning path table indicating if the student has started their learning path. Initially, the page shows a short description of how the learning path feature works. This message persists until the student starts the learning path manually.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- at least one competency with a lecture unit or Exercise connected

1. Log in to Artemis
2. Navigate to Learning Path feature
3. Verify that no learning path is present in the database
4. Start the learning path manually 
5. Verify that a learning path is present in the database with the startedByStudent flag set to true
6. Set the flag manually to false in the database
7. verify that the explanation text is still shown
8. Start the learning path manually
9. Verify that a learning path is present in the database with the startedByStudent flag set to true

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| metrics.model.ts | 100% | ✅ |
| learning-path-student-page.component.ts | 97.67% | ✅ ❌ |
| base-api-http.service.ts | 94.73% | ✅ ❌ |
| learning-path-api.service.ts | 100% | ✅ |
| learning-path.model.ts | 96.42% | ✅ ❌ |

### Screenshots

image

## Release Notes

- **New Features**
  - Introduced user-specific learning path management, allowing users to start and retrieve their learning paths.
  - Added a new `LearningPathDTO` for enhanced data representation of learning paths.

- **Improvements**
  - Updated API endpoints to provide comprehensive data, including the ability to start learning paths directly.
  - Enhanced user interface logic for better navigation and interaction with learning paths.

- **Bug Fixes**
  - Improved error handling and validation for learning path operations.

### CODE REVIEW

File: src/main/webapp/app/course/learning-paths/services/base-api-http.service.ts [hunk 1]
```diff
@@ -122,4 +122,34 @@ export abstract class BaseApiHttpService {
     ): Promise<T> {
         return await this.request<T>(HttpMethod.Post, url, { body: body, ...options });
     }
+
+    /**
+     * Constructs a `PUT` request that interprets the body as JSON and
```
Comment (pzdr7) [inline src/main/webapp/app/course/learning-paths/services/base-api-http.service.ts:127]: `PATCH`

File: src/main/webapp/app/course/learning-paths/services/base-api-http.service.ts [hunk 2]
```diff
@@ -122,4 +122,34 @@ export abstract class BaseApiHttpService {
...
+     * returns a Promise of an object of type `T`.
+     *
+     * @param url The endpoint URL excluding the base server url (/api).
+     * @param body The content to include in the body of the request.
+     * @param options The HTTP options to send with the request.
+     * @protected
+     *
+     * @return An `Promise` of type `Object` (T),
```
Comment (pzdr7) [inline src/main/webapp/app/course/learning-paths/services/base-api-http.service.ts:135]:
```suggestion
     * @return A `Promise` of type `Object` (T),
```

File: src/main/java/de/tum/cit/aet/artemis/atlas/service/learningpath/LearningPathService.java [hunk 3]
```diff
@@ -243,6 +245,52 @@ private void updateLearningPathProgress(@NotNull LearningPath learningPath) {
...
+     * @param learningPathId the id of the learning path
+     */
+    public void startLearningPathForCurrentUser(long learningPathId) {
+        final var learningPath = learningPathRepository.findByIdElseThrow(learningPathId);
+        final var currentUser = userRepository.getUser();
+        if (!learningPath.getUser().equals(currentUser)) {
+            throw new AccessForbiddenException("You are not allowed to start this learning path.");
+        }
+        else if (learningPath.isStartedByStudent()) {
+            throw new BadRequestException("Learning path already started.");
```
Comment (pzdr7) [inline src/main/java/de/tum/cit/aet/artemis/atlas/service/learningpath/LearningPathService.java:288]: Nitpick: Consider using a `ConflictException` here

File: src/main/webapp/app/course/learning-paths/pages/learning-path-student-page/learning-path-student-page.component.ts [hunk 4]
```diff
@@ -32,45 +32,49 @@ import { ArtemisSharedModule } from 'app/shared/shared.module';
...
     }
 
     private async loadLearningPathId(courseId: number): Promise<void> {
         try {
-            this.isLearningPathIdLoading.set(true);
-            const learningPathId = await this.learningApiService.getLearningPathId(courseId);
-            this.learningPathId.set(learningPathId);
+            this.isLearningPathLoading.set(true);
+            const learningPath = await this.learningApiService.getLearningPathForCurrentUser(courseId);
+            this.learningPath.set(learningPath);
```
Comment (pzdr7) [inline src/main/webapp/app/course/learning-paths/pages/learning-path-student-page/learning-path-student-page.component.ts:54]: This method loads the learning path now & not just its id. Can you rename it?

File: src/main/java/de/tum/cit/aet/artemis/atlas/service/learningpath/LearningPathService.java [hunk 5]
```diff
@@ -243,6 +245,52 @@ private void updateLearningPathProgress(@NotNull LearningPath learningPath) {
...
+        final var course = courseRepository.findWithEagerCompetenciesAndPrerequisitesByIdElseThrow(courseId);
+        if (learningPathRepository.findByCourseIdAndUserId(courseId, currentUser.getId()).isPresent()) {
+            throw new BadRequestException("Learning path already exists.");
+        }
+        final var learningPath = generateLearningPathForUser(course, currentUser);
+        return LearningPathDTO.of(learningPath);
+    }
+
+    /**
+     * Start the learning path for the current user in the given course.
```
Comment (pzdr7) [inline src/main/java/de/tum/cit/aet/artemis/atlas/service/learningpath/LearningPathService.java:277]: I don't think there's a "given course" here

```suggestion
     * Start the specified learning path for the current user.
```

**Review by pzdr7 — CHANGES_REQUESTED** [review]
A few small comments on the code. Please also check the failing server test (de.tum.cit.aet.artemis.atlas.learningpath.LearningPathIntegrationTest ► testAll_asStudent()) that fails because the corresponding endpoint is no longer available

File: src/main/java/de/tum/cit/aet/artemis/atlas/domain/competency/LearningPath.java [hunk 6]
```diff
@@ -31,6 +31,12 @@ public class LearningPath extends DomainObject {
     @Column(name = "progress")
     private int progress;
 
+    /**
+     * flag indicating if a student started the learning path
+     */
+    @Column(name = "started_by_student")
+    private boolean startedByStudent = false;
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/cit/aet/artemis/atlas/domain/competency/LearningPath.java:38]: Should we move this to the future LearnerProfile, that is stored once per student? Or should they get the explanation in each course?

File: src/main/resources/config/liquibase/master.xml [hunk 7]
```diff
@@ -23,7 +23,7 @@
     <include file="classpath:config/liquibase/changelog/20240708144500_changelog.xml" relativeToChangelogFile="false"/>
     <include file="classpath:config/liquibase/changelog/20240802091201_changelog.xml" relativeToChangelogFile="false"/>
     <include file="classpath:config/liquibase/changelog/20240804144500_changelog.xml" relativeToChangelogFile="false"/>
-
+    <include file="classpath:config/liquibase/changelog/20240924125742_changelog.xml" relativeToChangelogFile="false"/>
```
Comment (JohannesStoehr) [inline src/main/resources/config/liquibase/master.xml:26]: Can you re-add the empty line below all change files?

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

File: src/main/java/de/tum/cit/aet/artemis/atlas/domain/competency/LearningPath.java (hunk 6 above)
Comment (JohannesWt) [reply src/main/java/de/tum/cit/aet/artemis/atlas/domain/competency/LearningPath.java:38]: Makes sense for the future. @N0W0RK can you add that point to your backlog?

**Review by JohannesStoehr — APPROVED** [review]
Code

**Review by pzdr7 — APPROVED** [review]
Code

**Review by Jan-Thurner — APPROVED** [review]
tested locally in a testing session. Works as expected.

**Review by SimonEntholzer — APPROVED** [review]
Worked during testing :+1:

**Review by MaximilianAnzinger — APPROVED** [review]
Maintainer approved

**Pull request merged**